File: lab/1/HW1_9923085/main.py
import numpy as np


def sigmoid(x):
    """Sigmoid function"""
    return 2 / (1 + np.exp(-x)) - 1


def tanh(x):
    """tanh function"""
    return (np.exp(x) - np.exp(-x))/(np.exp(x) + np.exp(-x))


def sigmoid_derivative(x):
    """Derivative of sigmoid function"""
    sig = sigmoid(x)
    return (1 - sig**2)/2


def sigmatrix(matrix, r=None, c=None):
    """Function to process a matrix"""
    result1 = 0
    result2 = 0
    # Loop over each element in the matrix
    for row in matrix:
        for element in row:
            result1 += sigmoid(element)
            result2 += sigmoid_derivative(element)
    return (result1, result2)


def tanhmatrix(matrix):
    """Function to process a matrix"""
    result = 0
    for row in matrix:
        for element in row:
            result += tanh(element)
    return result


# Test the function with the given matrix
matrix = np.array([[1, 0, np.sin(np.pi / 4)],
                  [0, 1, np.sin(np.pi / 2)], [1, 0, 1]])
output1 = sigmatrix(matrix)
output2 = tanhmatrix(matrix)

print(output1)
print(output2)

File: lab/1/HW1_9923085/main.m
% Input the given matrix and calculate the result
matrix = [1, 0, sin(pi/4); 0, 1, sin(pi/2); 1, 0, 1];
output1 = sigmatrix(matrix,3,3);
output2 = tanhmatrix(matrix,3,3);
disp(['Result: ', num2str(output1)]);
disp(['Result: ', num2str(output2)]);

% Sigmoid function
function y = sigmoid(x)
    y = 2 / (1 + exp(-x)) - 1;
end

% tanh function
function y = tanh(x)
    y = (exp(x) - exp(-x))/(exp(x) + exp(-x));
end

% Derivative of sigmoid function
function y_prime = sigmoid_derivative(x)
    sig = sigmoid(x);
    y_prime = (1 - sig^2)/2;
end


% Function to process a matrix
function [res1, res2] = sigmatrix(matrix, rows, cols)
    res1 = 0;
    res2 = 0;
    % Loop over each element in the matrix
    for i = 1:rows
        for j = 1:cols
            res1 = res1 + sigmoid(matrix(i,j));
            res2 = res2 + sigmoid_derivative(matrix(i,j));
        end
    end
end

% Function to process a matrix
function res1 = tanhmatrix(matrix, rows, cols)
    res1 = 0;
    % Loop over each element in the matrix
    for i = 1:rows
        for j = 1:cols
            res1 = res1 + tanh(matrix(i,j));
        end
    end
end